training/controllers/api.py

```diff
@@ -0,0 +1,200 @@
+        return True
+    return False
+
+
+@bp.route('/register', methods=['POST'])
+def sign_up():
+    if not request.json or not 'username' in request.json or not 'password' in request.json or not 'email' in request.json:
+        abort(400)
+    else:
+        if repeated_fields(request.json):
+            abort(400)
+
+        id_user = request.json.get('username')
+        username = request.json.get('username')
+        password = bcrypt.generate_password_hash(request.json.get('password')).decode('utf-8')
+        email = request.json.get('email')
+
+        user = User(id=id_user, username=username, email=email, password=password)
+
+        db.session.add(user)
+        db.session.commit()
+
+        return jsonify({'username': username, 'password': password}), 201
+
+
+@bp.route('/login', methods=['POST'])
+def sign_in():
+    """
+    Important: here, an auth_token will be provided through the response body.
+    Then to access the endpoints that require a login (@login_required), you need to attach
+    the auth_token in the request header manually.
+    """
+
+    if not request.json or not 'username' in request.json or not 'password' in request.json:
+        abort(400)
+    else:
+        our_user = User.query.filter_by(id=request.json.get('username')).first()
+
+        if our_user is None:
+            abort(400)
+        else:
+            if our_user.mail_validation is False:
+                abort(401)
+            else:
+                if bcrypt.check_password_hash(our_user.password, request.json.get('password')):
+                    token = serializer.dumps(request.json.get('username'), salt='login')
+                    body = {
+                        "auth_token": token
+                    }
+                    return jsonify(body), 200
+                else:
+                    return "Bad password", 450
+
+
+@bp.route('/verify/email/<string:username>')
+def verify_email(username):
+    our_user = User.query.filter_by(id=username).first()
+
+    if our_user is None:
+        abort(404)
+
+    if our_user is None:
+        return "User not exist", 450
+    else:
+        if our_user.mail_validation:
+            return "The email {} was already validated.".format(our_user.username), 205
+        else:
+            our_user.mail_validation = True
+            db.session.commit()
+            return "Email from user {} was successfully validated.".format(our_user.username), 200
+
+
+@bp.route('/users/data/<string:username>')
+def return_user_data(username):
+    our_user = User.query.filter_by(id=username).first()
+
+    if our_user is None:
+        abort(404)
+
+    # user_json = json.dumps(our_user.__dict__)
+    user_json = {
+        "username": our_user.username,
+        "email": our_user.email,
+        "password": our_user.password,
+        "mail_validation": our_user.mail_validation,
+        "created_at": our_user.created_at
+    }
+
+    return user_json, 211
```

Comment:
What's the difference of returning a raw dict object and using `jsonify`?

Reply:
When you return something from a view functions. A Response Object is returned.
Response Object has three main attributes:
1. Body
2. Status code
3. Header

When you return just an string, the Mime/Type inside the header remains as default and flask returns a mime/type "text/html" and status code will be the default status code 200.
Jsonify returns a Response Object with a JSON as body and the correct mime/type automatically.